libmscore/mmrest.cpp

```diff
@@ -0,0 +1,214 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2020 MuseScore BVBA
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENCE.GPL
+//=============================================================================
+
+#include "mmrest.h"
+#include "measure.h"
+#include "score.h"
+#include "undo.h"
+#include "utils.h"
+
+namespace Ms {
+
+//---------------------------------------------------------
+//   mmRestStyle
+//---------------------------------------------------------
+
+static const ElementStyle mmRestStyle {
+      { Sid::mmRestNumberPos, Pid::MMREST_NUMBER_POS },
+      };
+
+//---------------------------------------------------------
+//    MMRest
+//--------------------------------------------------------
+
+MMRest::MMRest(Score* s)
+   : Rest(s)
+      {
+      _mmWidth   = 0;
+      if (score())
+            initElementStyle(&mmRestStyle);
+      }
+
+MMRest::MMRest(const MMRest& r, bool link)
+      {
+      if (link) {
+            score()->undo(new Link(this, const_cast<MMRest*>(&r)));
+            setAutoplace(true);
+            }
+      _mmWidth = r._mmWidth;
+      }
+
+//---------------------------------------------------------
+//   MMRest::draw
+//---------------------------------------------------------
+
+void MMRest::draw(QPainter* painter) const
+      {
+      if (shouldNotBeDrawn())
+            return;
+
+      qreal _spatium = spatium();
+      painter->setPen(curColor());
+
+      //only on voice 1
+      if (track() % VOICES)
+            return;
+
+      // draw number
+      int n = measure()->mmRestCount();
+      std::vector<SymId>&& s = toTimeSigString(QString("%1").arg(n));
+      qreal y = _mmRestNumberPos * spatium() - staff()->height() * .5;
+      qreal x = (_mmWidth - symBbox(s).width()) * .5;
+      drawSymbols(s, painter, QPointF(x, y));
+
+      // draw horizontal line
+      qreal pw = _spatium * .7;
+      QPen pen(painter->pen());
+      pen.setWidthF(pw);
+      painter->setPen(pen);
+      qreal x1 = pw * .5;
+      qreal x2 = _mmWidth - x1;
+      if (_mmRestNumberPos > 0.7 && _mmRestNumberPos < 3.3) { // hack for when number encounters horizontal line
```

Comment:
This will have to be cause conflict with PR #6105.

Reply:
Sure, if yours is merged first I will rebase.